File: Go/Easy/Generate Document/solution_2.go
package main

func GenerateDocument(characters string, document string) bool {

	countedCharacters := make(map[rune]int) // could be also map[rune]bool{}

	for _, character := range document {
		if countedCharacters[character] == int(character) {
			continue
		}

		documentFrequency := countFrequency(character, document)
		characterFrequency := countFrequency(character, characters)

		if documentFrequency > characterFrequency {
			return false
		}
	}

	return true

}

func countFrequency(targetCharacter rune, array string) int {

	var frequency = 0
	for _, char := range array {
		if char == targetCharacter {
			frequency++
		}
	}
	return frequency
}

File: Python/Easy/Generate Document/solution_2.py
def generateDocument(characters, document):

    countedCharacters = set()

    for character in document:

        if character in countedCharacters:
            continue

        documentFrequency = countFrequency(character, document)
        characterFrequency = countFrequency(character, characters)

        if documentFrequency > characterFrequency:
            return False
        
        countedCharacters.add(character)
            
    return True


def countFrequency(targetCharacter, array):
    frequency = 0
    for letter in array:
        if letter == targetCharacter:
            frequency += 1
    
    return frequency
